src-tauri/src/device/mesh_graph.rs

```diff
@@ -0,0 +1,402 @@
+use crate::graph::edge::GraphEdge;
+use crate::graph::graph_ds::Graph;
+use crate::graph::node::GraphNode;
+
+#[derive(Clone, Debug)]
+struct NodeTimeoutData {
+    timeout: Instant,
+}
+
+impl Eq for NodeTimeoutData {}
+
+impl PartialEq for NodeTimeoutData {
+    fn eq(&self, other: &Self) -> bool {
+        self.timeout == other.timeout
+    }
+}
+
+impl Ord for NodeTimeoutData {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // Inverted since we need to prioritize sooner (smaller) timeouts
+        // Lower number -> higher priority
+        other.timeout.cmp(&self.timeout)
+    }
+}
+
+impl PartialOrd for NodeTimeoutData {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+#[derive(Clone, Debug)]
+struct EdgeTimeoutData {
+    timeout: Instant,
+}
+
+impl Eq for EdgeTimeoutData {}
+
+impl PartialEq for EdgeTimeoutData {
+    fn eq(&self, other: &Self) -> bool {
+        self.timeout == other.timeout
+    }
+}
+
+impl Ord for EdgeTimeoutData {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // Inverted since we need to prioritize sooner (smaller) timeouts
+        // Lower number -> higher priority
+        other.timeout.cmp(&self.timeout)
+    }
+}
+
+impl PartialOrd for EdgeTimeoutData {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/*
+ * Just as the MeshDevice struct contains all the information about a device (in raw packet form)
+ * the MeshGraph struct contains the network info in raw graph form. This is synchronized with
+ * the MeshDevice struct, and is used to generate the graph visualization/algorithm
+ * results (see analytics).
+ */
+
+#[derive(Debug)]
+pub struct MeshGraph {
+    // TODO this shouldn't be public
+    pub graph: Graph,
+    node_timeout_queue: PriorityQueue<NodeIndex, NodeTimeoutData>,
+    edge_timeout_queue: PriorityQueue<EdgeIndex, EdgeTimeoutData>,
+}
+
+impl MeshGraph {
+    pub fn new() -> Self {
+        MeshGraph {
+            graph: Graph::new(),
+            node_timeout_queue: PriorityQueue::new(),
+            edge_timeout_queue: PriorityQueue::new(),
+        }
+    }
+
+    pub fn update_from_neighbor_info(&mut self, packet: NeighborInfoPacket) {
+        let NeighborInfoPacket { data, .. } = packet;
+        let protobufs::NeighborInfo {
+            neighbors,
+            node_broadcast_interval_secs,
+            ..
+        } = data.clone();
+
+        // Convert the NeighborInfo data into a GraphNode
+        let node_index = self.add_node(data.into());
+
+        for neighbor in neighbors {
+            // Implicitly convert the Neighbor data into a GraphNode
+            // let neighbor_index = self.add_node(neighbor.into());
+            let neighbor_index = self.add_node(GraphNode::new(
+                neighbor.node_id,
+                // TODO this should be used once the firmware doesn't zero the internal value
```

Comment:
I think the way to go is to loop through all neighbors that this node currently has, and remove those that are not in the newly incoming `NeighborInfoPacket`. That means they timed out in the firmware.